### ISSUE TITLE
Export internal Buffer statistics of logdrivers

### ISSUE THREAD

**Issue by daltonmatos** [moby/moby#37148]
Hello all,

Currently docker does not exposes any internal data about any of its logdrivers and their internal state, there's also currently no interface for this (in case a logdriver wants to publish these numbers).

The idea is to have the ability to publish (don't know how yet, maybe on the output of docker inspect?) some numbers about the internal state of the logdriver, such as:

* Is my internal buffer full?
* Am I having any difficulty delivering my logs to the configured destination?
* How many retries I already made?
* Total buffer size, in bytes;
* Buffer used space, in bytes;
* other useful metrics

With numbers like these it would be possible to monitor the health of the log collecting for each container, and whenever some problems arise to be able to identify where the problem is quicker.

What do you think about this possibility? The idea of this issue is to discuss about this and eventually get to a point where it's possible to implement.

Thanks,

**Comment by thaJeztah** [comment]
perhaps in the prometheus metrics?

@cpuguy83 wdyt?

**Comment by daltonmatos** [comment]
Hello @thaJeztah,

Indeed this could be a good solution, and much better that docker inspect. =D

Does docker currently exposes any metrics that prometheus can scrape? Are there any API endpoints do expose these? Maybe this could be a good solution. A new "`/metrics`" endpoint on the docker API that plugins can add data and that can be read by anyone interested, including prometheus.

Let's see what other devs think about this.

Thanks,

**Comment by thaJeztah** [comment]
Yes, metrics are already exposed, see https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-metrics

**Comment by daltonmatos** [comment]
That's awesome @thaJeztah, didn't know about. Will take a look. Thank you very much!

Are there any docs about which metrics are currently being exported?

**Comment by thaJeztah** [comment]
> Are there any docs about which metrics are currently being exported?

No there's no documentation currently; reason for that was that the names and type of metrics were not finalized, and we wanted to be able to  change them, based on feedback from people trying the metrics

More info on that can be found in https://github.com/moby/moby/issues/27307, https://github.com/moby/moby/issues/31107, https://github.com/moby/moby/issues/32948

**Comment by daltonmatos** [comment]
Hello @cpuguy83,

> We did just add metrics for errors when any log driver reports an error when delivering logs, but this is global.

Is this shipped in some version of docker or is it still only in the dev version? I'm running docker `18.05.0-ce` and it seem not to have it yet. Looking at the output of `/metrics` I didn't see any metric related to logs.

Thanks!

**Comment by cpuguy83** [comment]
Not shipped yet.

**Comment by daltonmatos** [comment]
OK @cpuguy83, thanks for the info. I will try to build this dev version and take a look at this new metric.


> A bit worried to expose this per container because prometheus really doesn't like labels for an unbounded set as it increases storage requirements... but maybe someone more familiar with metrics can speak to that.

Who else you think we should mention here so they can express their opinion about this possible detailed-logdriver-per-container metrcis?

> do you think that would be a good idea to be able to enable/disable the desired metrics? Doing this would make it possible to opt-in to a "bigger prometheus" in exchange for more detailed metrics.
>
> Maybe some more --metrics-* options to do this.

What would you think about this approach?

**Comment by daltonmatos** [comment]
Hello @thaJeztah,

anyone else we could mention here?

Thanks,
